### PR TITLE
#467 VS hangs after closing NTVS project

### LINKED ISSUE #467: VS2015 hangs after closing NTVS project
Hi,
First off, I want to say that NTVS is simply amazing! It's the most advanced way to develop and debug Node.js apps as of today. The IntelliSense is so nice to have! And so is the "edit-and-continue". No other IDE is able to match these features, and I tested a lot of them. IntelliSense for JS is extremely hard to pull off. Well done.

One thing that happens randomly is VS2015 freezing when closing a NTVS solution. It happens about 80% of the time. Everything works great until I'm done with the project and want to close it. VS will hang and won't exit even after minutes. Only a process kill via Task Manager to `devenv.exe` can close VS.

Is this a known issue? I tried searching and couldn't find it.

Thanks for your support,
Elad

### LINKED ISSUE #505: Hang while populating completion list
**app.js**
```js
var c= requestId.// long hang after pressing .
```

**package.json:**
```json
{
  "name": "App",
  "main": "app.js",
  "version": "1.0.0",
  "dependencies": {
    "async": "^0.9.0",
    "body-parser": "^1.10.1",
    "cookie-parser": "^1.3.4",
    "crypto": "0.0.3",
    "express": "~4.9.0",
    "mongoose": "^4.1.7",
    "mqtt": "^0.3.13",
    "nodemailer": "^1.3.4",
    "nodemailer-ses-transport": "^1.2.0",
    "path": "^0.11.14",
    "request": "^2.51.0",
    "stripe": "^3.3.4",
    "validator": "^4.0.5"
  },
  "devDependencies": {
    "jshint": "^2.8.0"
  }
}
````

Additionally, it looks like way too many types are flowing through here:
image

### PR DESCRIPTION
**Pull request by mousetraps**
#505 Hang while populating completion list
- we weren't respecting MaxMergeTypes everywhere during analysis
- also fix InvalidCastException in GlobalBuilder.Require

(I found some other places where this may be an issue, but I'm not 100% sure, so I'll have to ask Dino about them verify.)

fix #467, fix #505

### CODE REVIEW

File: Nodejs/Product/Analysis/Analysis/GlobalBuilder.cs [hunk 1]
```diff
@@ -1310,7 +1310,8 @@ class GlobalBuilder {
 
         private static IAnalysisSet DefineProperties(FunctionValue func, Node node, AnalysisUnit unit, IAnalysisSet @this, IAnalysisSet[] args) {
             // object, {propName: {desc}, ...}
-            if (args.Length >= 2) {
+            var maxMergeTypes = unit.Analyzer.Limits.MaxMergeTypes;
```
Comment (billti) [inline Nodejs/Product/Analysis/Analysis/GlobalBuilder.cs]: I assume MaxMergeTypes always has some positive integer then - regardless of analysis settings?

**Comment by billti** [comment]
One question, else based on my limited knowledge :+1:

File: Nodejs/Product/Analysis/Analysis/GlobalBuilder.cs (hunk 1 above)
Comment (mousetraps) [inline Nodejs/Product/Analysis/Analysis/GlobalBuilder.cs]: yep - in fact, this value is currently the same with every setting.

File: Nodejs/Product/Analysis/Analysis/GlobalBuilder.cs [hunk 2]
```diff
@@ -1310,7 +1310,8 @@ class GlobalBuilder {
...
+            if (args.Length >= 2 && args[0].Count < maxMergeTypes && args[1].Count < maxMergeTypes) {
```
Comment (mousetraps) [inline Nodejs/Product/Analysis/Analysis/GlobalBuilder.cs]: Only check args[0] here

**Pull request merged**